src/core/config/transformer/flatten_single_types.rs

```diff
@@ -0,0 +1,176 @@
+use std::collections::BTreeMap;
+
+use crate::core::config::{Config, Field, Type};
+use crate::core::transform::{self, Transform};
+use crate::core::valid::Valid;
+
+/// Flatten single types
+/// If a type has only one field, it is considered a single type.
+/// If a field references a single type, it will be flattened.
+#[derive(Default)]
+pub struct FlattenSingleTypes;
+
+impl Transform for FlattenSingleTypes {
+    type Value = Config;
+    type Error = String;
+
+    fn transform(&self, mut config: Config) -> Valid<Config, String> {
+        let single_types = get_single_types(&config);
+        let root_types = get_root_types(&config);
+        let mut still_referenced_single_types = Vec::<String>::new();
+
+        config.types.iter_mut().for_each(|(_, ty)| {
+            let fields_referencing_single_type = ty
+                .fields
+                .iter_mut()
+                .filter(|(_, field)| single_types.contains_key(&field.type_of));
+
```

Comment:
I did skip the flatten when there is a resolver cause i was anticipating some side effects.